### PR TITLE
withTooltip HOC

### LINKED ISSUE #599: Create withTooltip HOC
Let's create a HOC to handle the Tooltip visibility.

### PR DESCRIPTION
**Pull request by fgameiro**
affects: @crave/farmblocks-hoc-withtooltip

## Closes Issues

* Closes #599 

### Reviewer 1

* [ ] Functions properly
* [ ] Good readability, clarity and overall quality
* [ ] Has tests
* [ ] Has docs

### Reviewer 2

* [ ] Functions properly
* [ ] Good readability, clarity and overall quality
* [ ] Has tests
* [ ] Has docs

### CODE REVIEW

File: packages/hoc-disabled-tooltip/src/disabledTooltip.js [hunk 1]
```diff
@@ -19,50 +19,30 @@ function disabledTooltip(wrappedComponentType) {
...
-              <div>
-                <Tooltip text={tooltipText} align={tooltipAlign} />
-              </div>
-            )}
+          <WrappedComponentWithTooltip
+            onMouseOver={onMouseOver}
+            onMouseLeave={onMouseLeave}
+            disableTooltip={!this.props.disabled || noTooltip}
+            tooltipText={tooltipText}
+            {...wrappedComponentProps}
```
Comment (vNakamura) [inline packages/hoc-disabled-tooltip/src/disabledTooltip.js]: `onMouseOver`, ` onMouseLeave`, `tooltipAlign` and `tooltipText` could be part of the `wrappedComponentProps`. No need to extract from props and then passed back to the wrapped component with the same names.

File: packages/hoc-withtooltip/src/withTooltip.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import * as React from "react";
+import Tooltip, { alignments } from "@crave/farmblocks-tooltip";
+import wrapDisplayName from "recompose/wrapDisplayName";
+import PropTypes from "prop-types";
+import styled from "styled-components";
+import values from "object.values";
+
+const EnhancedDiv = styled.div`
+  position: relative;
+  display: inline-block;
```
Comment (vNakamura) [inline packages/hoc-withtooltip/src/withTooltip.js]: The `enhanced element with display block` story from `hoc-disabled-tooltip` is no longer working as expected.
Before:

After:


Take a look at the `Wrapper` styled component from the hoc-disabled. You can move the `displayBlock` prop to here and remove the whole wrapper from there.

File: packages/hoc-disabled-tooltip/src/disabledTooltip.js [hunk 3]
```diff
@@ -19,50 +19,30 @@ function disabledTooltip(wrappedComponentType) {
...
-      return this.props.onMouseOver(event);
-    }
-
-    onMouseLeave(event) {
-      this.setState({ showTooltip: false });
-      return this.props.onMouseLeave(event);
-    }
-
     render() {
+      const WrappedComponentWithTooltip = withTooltip(WrappedComponent);
```
Comment (alcferreira) [inline packages/hoc-disabled-tooltip/src/disabledTooltip.js]: It's composing the components on each render cycle... Let's remove this from here.

File: packages/hoc-withtooltip/src/withTooltip.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+          </Tooltip>
+        </Container>
+      );
+    }
+
+    static displayName = wrapDisplayName(WrappedComponent, "withTooltip");
+
+    static propTypes = {
+      disableTooltip: PropTypes.bool,
+      tooltipContent: PropTypes.oneOfType(PropTypes.string, PropTypes.node),
```
Comment (alcferreira) [inline packages/hoc-withtooltip/src/withTooltip.js]:
```suggestion
      tooltipContent: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
```

File: packages/hoc-withtooltip/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+```
+import React, { Component } from 'react';
+import { render } from 'react-dom';
+import withTooltip from '@crave/farmblocks-hoc-withTooltip'
+
+const InputWithTooltip = withTooltip("input");
+
+const App = props => (
+  <div>
+    <InputWithTooltip tooltipText="This is the text rendered inside the tooltip" />
```
Comment (alcferreira) [inline packages/hoc-withtooltip/README.md]: `tooltipText` has changed to `tooltipContent `.

Readme should point out that this hoc uses the tooltip component and which components we are expecting

File: packages/hoc-withtooltip/src/withTooltip.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+const Container = styled.div`
+  position: relative;
+  display: ${props => (props.displayBlock ? "block" : "inline-block")};
+
+  .hitArea {
+    position: absolute;
+    top: 0;
+    width: 100%;
+    height: 100%;
```
Comment (vNakamura) [inline packages/hoc-withtooltip/src/withTooltip.js]: This element is covering the wrapped component. Click events and hover styles are not been triggered.

Check the `HOC disabledTooltip - disabled false and button events (click, mouseover)` story.

**Pull request merged**